pulp_ansible/app/migrations/0047_add_repo_cv_index.py

```diff
@@ -0,0 +1,144 @@
+#
+# Why?
+#   * The x-repo search requirements want the same CV listed N times per
+#     repositories containing the CV.
+#   * Many models and objects and concepts desired for cross repo search
+#     filtering are not directly related to repositories.
+#   * Using the queryset language to build the counts for signatures and
+#     deprecations were drifting from wizardry into impossibility.
+#   * Creating a real index model requires too many hooks in various
+#     obscure workflows for CRUD and new devs will always make mistakes.
+
+CREATE_REPOSITORY_COLLECTIONVERSION_VIEW = '''
+CREATE OR REPLACE VIEW repository_collection_version AS
+
+-- make a map for highest version
+WITH HighestVersionLookup AS
+(
+    SELECT
+        MAX(number) as highest_version,
+        repository_id as highest_version_repository_id
+    FROM
+        core_repositoryversion
+    WHERE
+        number>0
+    GROUP BY
+        repository_id
+    ORDER BY
+        repository_id
+)
+
+SELECT
+
+    crc.pulp_id as crc_pulp_id,
+    crc.content_id as crc_content_id,
+    crc.repository_id as crc_repository_id,
+    crv.number as crv_version_number,
+    crc.version_added_id as crc_version_added_id,
+    crc.version_removed_id as crc_version_removed_id,
+    cr.pulp_id as cr_pulp_id,
+
+    cr.next_version as cr_next_version,
+    acv.content_ptr_id as collectionversion_id,
+
+    hvl.highest_version_repository_id as highest_version_repository_id,
+    hvl.highest_version as highest_version,
+
+    -- make a synthetic ID for uniqueness
+    concat(cr.pulp_id, ':', acv.content_ptr_id) AS id,
+
+    -- used by the model
+    cr.pulp_id as repository_id,
+
+    -- used by the model
+    cr.name as reponame,
+
+    -- used by the model
+    acv.namespace as namespace,
+
+    -- used by the model
+    acv.name as name,
+
+    -- used by the model
+    acv.version as version,
+
+    -- count the number of deprecations for this namespace.name
+    (
+        SELECT
+            COUNT(*)
+        FROM
+            ansible_ansiblecollectiondeprecated acd
+        WHERE
+            acd.namespace=acv.namespace
+        AND
+            acd.name=acv.name
+    ) as deprecation_count,
+
+    -- count the number of applicable signatures for this CV+repo
+    (
+        SELECT
+            COUNT(*)
+        FROM
+            ansible_collectionversionsignature acvs
+        WHERE
+            acvs.signed_collection_id=acv.content_ptr_id
+        AND
+        (
+            SELECT
+                COUNT(*)
+            FROM
+                core_repositorycontent crc3
+            WHERE
+                crc3.repository_id=crc.repository_id
+            AND
+                crc3.content_id=acvs.content_ptr_id
+        )>=1
+    ) as sig_count
+
+FROM
+    core_repositorycontent crc
+INNER JOIN
+    core_repository cr ON cr.pulp_id=crc.repository_id
+INNER JOIN
+    core_repositoryversion crv ON crv.repository_id=cr.pulp_id
```

Comment:
How does this INNER JOIN work? Again normal pulp with full repo history can have multiple repo versions pointing to a repository, so I don't see how this join would work in that case, but maybe I'm missing something.

Reply:
The inner join on core_repositoryversion (crv) is joining the core_repositorycontent (crc) table with the core_repositoryversion table on the repository_id. This means that for each row in the core_repositorycontent table, the query will only include the corresponding rows from the core_repositoryversion table that have the same repository_id. This join operation allows the query to retrieve the version number of the repository which the collection version is in.

Reply:
@gerrod3 maybe this will help break it down https://gist.github.com/jctanner/c3a77883aa957de9a0e50be7fc4a3811